### PR TITLE
Add set-label function

### PR DESCRIPTION
**Pull request by Shell32-Natsu**

### CODE REVIEW

**Comment by Shell32-Natsu** [comment]
cc @mengqiy

**Comment by Shell32-Natsu** [comment]
CI failed due to Go package 4d63.com/gochecknoglobals/checknoglobals not available

**Comment by frankfarzan** [comment]
I'll review this after https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/123 is approved since many comments will be applicable to both.

**Comment by Shell32-Natsu** [comment]
@frankfarzan Ready for review

**Comment by Shell32-Natsu** [comment]
e2e test is now using new driver.

@mengqiy @frankfarzan

File: functions/go/set-label/LabelTransformer_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
+package main
```
Comment (mengqiy) [inline functions/go/set-label/LabelTransformer_test.go:1]: Are these tests generated from the tests in kustomize or manually written?

File: functions/go/set-label/main.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+Example:
+
+To add a label 'color: orange' to all resources:
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config
+data:
+  label_name: color
```
Comment (mengqiy) [inline functions/go/set-label/main.go:105]: Do we support adding more than 1 labels at the same time?

**Review by mengqiy — COMMENTED** [review]
Overall looks good.

File: functions/go/set-label/LabelTransformer_test.go (hunk 1 above)
Comment (Shell32-Natsu) [reply functions/go/set-label/LabelTransformer_test.go:1]: I added them manually to avoid importing bunch of kustomize internal test utilities.

File: functions/go/set-label/main.go (hunk 2 above)
Comment (Shell32-Natsu) [reply functions/go/set-label/main.go:105]: No, this function only supports one label a time.

Comment (mengqiy) [reply functions/go/set-label/main.go:105]: Comparing with the functionality in kustomize, user can specify multiple labels under `commonLabels` in `kustomization.yaml`. 
When a kustomize user tries to use this function, he/she may expect it to be able to change multiple labels at once.

We need to decide this for set-label and set-annotation. @frankfarzan WDYT?

If we want to do multiple, we may have something like this:
```
type functionConfigData struct {
	Labels  []label `json:"labels,omitempty" yaml:"labels,omitempty"`
}

type label struct {
	LabelName  string `json:"label_name,omitempty" yaml:"label_name,omitempty"`
	LabelValue string `json:"label_value,omitempty" yaml:"label_value,omitempty"`
}
```

Comment (Shell32-Natsu) [reply functions/go/set-label/main.go]: Updated the PR to support multiple labels.

File: examples/set-label/functions/fn-config.yaml [hunk 3]
```diff
@@ -8,5 +8,12 @@ metadata:
...
     config.kubernetes.io/local-config: 'true'
 data:
-  label_name: color
-  label_value: orange
+  specs:
+  - label_name: color
+    label_value: orange
+  - label_name: fruit
+    label_value: apple
+    fieldSpecs:
```
Comment (mengqiy) [inline examples/set-label/functions/fn-config.yaml:16]: It seems this is per-label config. Why? Should it apply to all of the labels?

File: examples/set-label/functions/fn-config.yaml [hunk 4]
```diff
@@ -8,5 +8,12 @@ metadata:
         image: gcr.io/kpt-functions/set-label
     config.kubernetes.io/local-config: 'true'
 data:
-  label_name: color
-  label_value: orange
+  specs:
```
Comment (mengqiy) [inline examples/set-label/functions/fn-config.yaml:11]: I saw you use `rn.Pipe(kyaml.Lookup("data", "specs"))` below.
Why don't we define a corresponding go struct and simply use unmarshal? IMO the latter approach is easier to reasonable about.

File: examples/set-label/functions/fn-config.yaml (hunk 3 above)
Comment (Shell32-Natsu) [reply examples/set-label/functions/fn-config.yaml:16]: The default config is applied to all labels. Each label can have it's own custom spec.

File: examples/set-label/functions/fn-config.yaml (hunk 4 above)
Comment (Shell32-Natsu) [reply examples/set-label/functions/fn-config.yaml:11]: I don't think we should implement our own struct for function config given this part may be used by other functions and they have already had well-defined utilities in kyaml.

File: examples/set-label/functions/fn-config.yaml (hunk 3 above)
Comment (mengqiy) [reply examples/set-label/functions/fn-config.yaml:16]: I see. Thanks for clarification.

File: examples/set-label/functions/fn-config.yaml (hunk 4 above)
Comment (mengqiy) [reply examples/set-label/functions/fn-config.yaml:11]: Sorry for not being very clear.
I meant `data` is standard field for `ConfigMap`, but `data.specs` is not. `kyaml.Lookup("data", "specs")` this feels like hardcoding the path.
IMO it's better to treat whatever under `.data` as a blob and unmarshal the whole blob to `addLabelSpecs`.

`addLabelSpecs` may look like this:
```
type addLabelSpecs struct {
	Labels []addLabelSpec `json:"labels,omitempty" yaml:"labels,omitempty"`
	FieldSpec types.FieldSpec `json:"fieldSpec,omitempty" yaml:"fieldSpec,omitempty"` // Field config for all labels
}

type addLabelSpec struct { // you already defined this in functions/go/set-label/main.go
	LabelName  string            `json:"label_name,omitempty" yaml:"label_name,omitempty"`
	LabelValue string            `json:"label_value,omitempty" yaml:"label_value,omitempty"`
	FieldSpecs []types.FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"` // Field config for one label.
}
```
Then you have something like:
```
d := kyaml.Lookup("data")
yaml.Unmarshal(d, &ads)
```

Does it make sense to you?

Comment (Shell32-Natsu) [reply examples/set-label/functions/fn-config.yaml]: done

**Review by mengqiy — APPROVED** [review]
LGTM

**Comment by mengqiy** [comment]
Let's wait for some time to see if Frank want to take a look.

File: examples/set-label/functions/fn-config.yaml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  annotations:
+    config.k8s.io/function: |
+      container:
+        image: gcr.io/kpt-functions/set-label
+    config.kubernetes.io/local-config: "true"
+data:
+  labels:
+  - label_name: color
+    label_value: orange
+  - fieldSpecs:
```
Comment (frankfarzan) [inline examples/set-label/functions/fn-config.yaml:14]: @droot We need to rationalize these with `filters` at the Pipeline level.

@Shell32-Natsu @mengqiy For context, instead of each function implementing this filtering mechanism, we can provide this at a higher level in the `Pipeline` which makes function implementation simpler.

File: examples/set-label/.expected/diff.patch [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
++++ b/configs/example-config.yaml
+@@ -2,7 +2,11 @@ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: the-map
++  labels:
++    color: orange
++    fruit: apple
+ data:
++  fruit: apple
```
Comment (frankfarzan) [inline examples/set-label/.expected/diff.patch:13]: what's this being changed?

File: examples/set-label/.expected/diff.patch [hunk 7]
```diff
@@ -0,0 +1,55 @@
+diff --git a/configs/example-config.yaml b/configs/example-config.yaml
+index 1c43349..02e8ab4 100644
+--- a/configs/example-config.yaml
++++ b/configs/example-config.yaml
+@@ -2,7 +2,11 @@ apiVersion: v1
+ kind: ConfigMap
+ metadata:
+   name: the-map
++  labels:
```
Comment (frankfarzan) [inline examples/set-label/.expected/diff.patch:9]: Do tests currently check for idempotency?

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: examples/set-label/.expected/diff.patch (hunk 6 above)
Comment (Shell32-Natsu) [reply examples/set-label/.expected/diff.patch:13]: Because there is a custom field spec for this.

File: examples/set-label/.expected/diff.patch (hunk 7 above)
Comment (Shell32-Natsu) [reply examples/set-label/.expected/diff.patch:9]: The test driver doesn't check it. However, the unit test covers the idempotency of the transformer and I have updated the tests to 
```
  - label_name: color
    label_value: orange
  - label_name: color
    label_value: orange
  - fieldSpecs:
    - kind: ConfigMap
      create: true
      path: data
    label_name: fruit
    label_value: apple
```

File: examples/set-label/.expected/diff.patch (hunk 6 above)
Comment (frankfarzan) [reply examples/set-label/.expected/diff.patch:13]: Not following, what is a function for setting labels changing fields that are not labels?

File: examples/set-label/.expected/diff.patch (hunk 7 above)
Comment (frankfarzan) [reply examples/set-label/.expected/diff.patch:9]: @mengqiy This should be done by the e2e test harness.

File: examples/set-label/.expected/diff.patch (hunk 6 above)
Comment (Shell32-Natsu) [reply examples/set-label/.expected/diff.patch:13]: This is used to test the custom field specs. Users can use it to add labels to any field in any resource.

File: examples/set-label/.expected/diff.patch (hunk 7 above)
Comment (Shell32-Natsu) [reply examples/set-label/.expected/diff.patch:9]: https://github.com/GoogleContainerTools/kpt/issues/1396
 Will do it. But that doesn't block this PR.

Comment (mengqiy) [reply examples/set-label/.expected/diff.patch:9]: Yes, our test driver should cover idempotency.
Thanks for bring this up!

File: examples/set-label/functions/fn-config.yaml (hunk 5 above)
Comment (droot) [reply examples/set-label/functions/fn-config.yaml:14]: After looking at this and offline discussions, here is the summary:
`fieldSpecs` provided here is the schema information that `set-label` needs to support different types. So this completely orthogonal to the `filtering` mechanism in the pipeline.

Another bit that confused me is that currently schema and input args (label/value in this case) is mixed together that makes the usage look very different from the case where schema is not provided, so there is an opportunity to unify that as well. 

so, TL;DR is we need a mechanism to make schema information available to all KRM functions in an easy way. And while doing so, we should also unify the UX for taking input for the functions with or without schema info.

Comment (frankfarzan) [reply examples/set-label/functions/fn-config.yaml:14]: To expand on this, I consider the `fieldSpec` to be a low-level crutch until we address the general problem of augmenting the OpenAPI document with such semantics. A general solution would address all the following:

- Which field in a type is a `label`?
- Where is a field references from other resources?
- etc.

Kustomize has hardcoded these for k8s native types, we need to generalize this to address other resources types, e.g. Cloud Connector types.

File: examples/set-label/.expected/diff.patch (hunk 6 above)
Comment (frankfarzan) [reply examples/set-label/.expected/diff.patch:13]: Even though this is technically true, I think saying that ConfigMap.data is a label creates confusion. Instead use a custom resource with a label field.

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: examples/set-label/.expected/diff.patch (hunk 6 above)
Comment (Shell32-Natsu) [reply examples/set-label/.expected/diff.patch]: Updated the test case.

**Review by frankfarzan — APPROVED** [review]

**Comment by frankfarzan** [comment]
Some post-commit comments:
- Go files should use c_style_naming and not CamelCase naming
- Optional: Consider having directory organization for each function. Currently, everything is flat. `TS` functions have `build` to contain `Dockerfile` and `src` for sources. For Golang, we may need to also have `build` for `Dockerfile` and `pkg` for source files. I don't feel strongly about this though.

**Comment by Shell32-Natsu** [comment]
Will update go file names in separate pr
